homeassistant/components/tesla/__init__.py

```diff
@@ -34,69 +40,150 @@
+            if email == entry.title:
+                hass.config_entries.async_update_entry(
+                    entry, data=data, options=options
+                )
+
+    config = base_config.get(DOMAIN)
+    if not config:
+        return True
     email = config.get(CONF_USERNAME)
     password = config.get(CONF_PASSWORD)
-    update_interval = config.get(CONF_SCAN_INTERVAL)
-    if hass.data.get(DOMAIN) is None:
+    scan_interval = config.get(CONF_SCAN_INTERVAL, 300)
+    if email in configured_instances(hass):
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if email == entry.title:
+                try:
+                    websession = aiohttp_client.async_get_clientsession(hass)
+                    controller = teslaAPI(
+                        websession, email=email, password=password, update_interval=300
+                    )
+                    (refresh_token, access_token) = await controller.connect(
+                        test_login=True
+                    )
+                    hass.config_entries.async_update_entry(
+                        entry,
+                        data={
+                            CONF_ACCESS_TOKEN: access_token,
+                            CONF_TOKEN: refresh_token,
+                        },
+                        options={CONF_SCAN_INTERVAL: scan_interval},
+                    )
+                except TeslaException as ex:
+                    _LOGGER.warning(
+                        "Configuration.yaml credentials cannot communicate with Tesla API: %s",
+                        ex.message,
+                    )
+                break
+    else:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data={CONF_USERNAME: email, CONF_PASSWORD: password},
+            )
+        )
+        async_call_later(hass, 15, lambda _: _update_entry(email, scan_interval))
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up Tesla as config entry."""
+
+    if DOMAIN not in hass.data or config_entry.entry_id not in hass.data[DOMAIN]:
+        config = config_entry.data
+        if DOMAIN not in hass.data:
+            hass.data[DOMAIN] = {}
         try:
             websession = aiohttp_client.async_get_clientsession(hass)
             controller = teslaAPI(
                 websession,
-                email=email,
-                password=password,
-                update_interval=update_interval,
+                refresh_token=config[CONF_TOKEN],
+                update_interval=config_entry.options.get(CONF_SCAN_INTERVAL, 300),
             )
-            await controller.connect(test_login=False)
-            hass.data[DOMAIN] = {"controller": controller, "devices": defaultdict(list)}
+            (refresh_token, access_token) = await controller.connect()
```

Comment:
Does the refresh token change when you connect with a refresh token?

Reply:
Yes.